File: packages/ts-sdk/src/task.ts
import { randomUUID } from 'node:crypto'
import { withRunLedgerContext } from './context.js'
import type { Transport } from './transport.js'
import type { TaskOptions, RunEndEvent, RunStartEvent, ToolCallEvent, ProviderCallEvent, OutcomeEvent, SpanStartEvent, SpanEndEvent } from './types.js'

export class RunLedgerTask {
  readonly runId: string
  private _outcomeRecorded = false
  private _totalCostUsd = 0
  private _totalInputTokens = 0
  private _totalOutputTokens = 0

  constructor(
    private readonly _transport: Transport,
    runId: string,
    readonly name: string,
  ) {
    this.runId = runId
  }

  recordSpan(name: string, opts: { spanType?: 'llm' | 'tool' | 'retrieval'; status?: 'succeeded' | 'failed'; metadata?: Record<string, unknown> } = {}): string {
    const spanId = randomUUID()
    const now = new Date().toISOString()
    const start: SpanStartEvent = {
      event_type: 'span_start',
      run_id: this.runId,
      span_id: spanId,
      span_type: opts.spanType ?? 'tool',
      name,
      started_at: now,
    }
    const end: SpanEndEvent = {
      event_type: 'span_end',
      run_id: this.runId,
      span_id: spanId,
      status: opts.status ?? 'succeeded',
      ended_at: now,
      metadata: opts.metadata,
    }
    this._transport.enqueue(start)
    this._transport.enqueue(end)
    return spanId
  }

  recordToolCall(toolName: string, opts: Omit<ToolCallEvent, 'event_type' | 'run_id' | 'tool_name'> = { status: 'success' }): void {
    this._transport.enqueue({
      event_type: 'tool_call',
      run_id: this.runId,
      tool_name: toolName,
      ...opts,
    })
  }

  recordModelCall(opts: Omit<ProviderCallEvent, 'event_type' | 'run_id'>): void {
    this._transport.enqueue({
      event_type: 'provider_call',
      run_id: this.runId,
      ...opts,
    })
    this._totalCostUsd += Number(opts.cost_usd ?? 0)
    this._totalInputTokens += Number(opts.input_tokens ?? 0)
    this._totalOutputTokens += Number(opts.output_tokens ?? 0)
  }

  recordOutcome(
    outcomeType: string,
    opts: {
      success?: boolean
      labels?: Record<string, unknown>
      finalStatus?: 'succeeded' | 'failed'
      totalCostUsd?: number
      totalInputTokens?: number
      totalOutputTokens?: number
      qualityScore?: number
      verificationStatus?: string
    } = {},
  ): void {
    const labels = { ...(opts.labels ?? {}) }
    if (opts.qualityScore !== undefined) labels['quality_score'] = opts.qualityScore
    if (opts.verificationStatus !== undefined) labels['verification_status'] = opts.verificationStatus
    const outcome: OutcomeEvent = {
      event_type: 'outcome',
      run_id: this.runId,
      outcome_type: outcomeType,
      success: opts.success ?? true,
      labels: Object.keys(labels).length ? labels : undefined,
    }
    const end: RunEndEvent = {
      event_type: 'run_end',
      run_id: this.runId,
      ended_at: new Date().toISOString(),
      status: opts.finalStatus ?? 'succeeded',
      total_input_tokens: opts.totalInputTokens ?? this._totalInputTokens,
      total_output_tokens: opts.totalOutputTokens ?? this._totalOutputTokens,
    }
    if (opts.totalCostUsd !== undefined || this._totalCostUsd > 0) {
      ;(end as RunEndEvent & { total_cost_usd?: number }).total_cost_usd = opts.totalCostUsd ?? this._totalCostUsd
    }
    this._transport.enqueue(outcome)
    this._transport.enqueue(end)
    this._outcomeRecorded = true
  }

  complete(): void {
    this.recordOutcome('completed')
  }

  fail(error: unknown): void {
    const errorType = error instanceof Error ? error.name : 'Error'
    this.recordOutcome('failed', {
      success: false,
      finalStatus: 'failed',
      labels: { error_type: errorType },
    })
  }

  get outcomeRecorded(): boolean {
    return this._outcomeRecorded
  }
}

export async function withTask<T>(
  transport: Transport,
  name: string,
  options: TaskOptions,
  defaultMetadata: Record<string, unknown>,
  fn: (task: RunLedgerTask) => Promise<T> | T,
): Promise<T> {
  return withRunLedgerContext(options, async (runId) => {
    const mergedMetadata = { ...defaultMetadata, ...(options.metadata ?? {}), task: name }
    const start: RunStartEvent = {
      event_type: 'run_start',
      run_id: runId,
      started_at: new Date().toISOString(),
      end_user_id: options.endUserId,
      session_id: options.sessionId,
      feature_tag: options.featureTag,
      deployment_version: options.deploymentVersion,
      metadata: mergedMetadata,
      intent: options.intent,
    }
    transport.enqueue(start)
    const task = new RunLedgerTask(transport, runId, name)
    try {
      const result = await fn(task)
      if (!task.outcomeRecorded) task.complete()
      return result
    } catch (error) {
      if (!task.outcomeRecorded) task.fail(error)
      throw error
    }
  })
}

File: packages/sdk/runledger_sdk/task.py
"""High-level task wrapper for manual agent lifecycle instrumentation."""

from __future__ import annotations

from decimal import Decimal
from typing import Any


class RunLedgerTask:
    """Context manager that wraps a single task run with helper methods."""

    def __init__(
        self,
        client: Any,
        name: str,
        *,
        intent: str | None = None,
        metadata: dict[str, Any] | None = None,
        end_user_id: str | None = None,
        session_id: str | None = None,
        feature_tag: str | None = None,
        deployment_version: str | None = None,
    ) -> None:
        self._client = client
        self.name = name
        self.intent = intent
        self.metadata = metadata or {}
        self.end_user_id = end_user_id
        self.session_id = session_id
        self.feature_tag = feature_tag
        self.deployment_version = deployment_version
        self.run_id: str | None = None
        self._ctx: Any | None = None
        self._outcome_recorded = False
        self._total_cost_usd = Decimal("0")
        self._total_input_tokens = 0
        self._total_output_tokens = 0

    def __enter__(self) -> RunLedgerTask:
        default_metadata = getattr(self._client, "_default_task_metadata", {})
        merged_metadata = {**default_metadata, **self.metadata, "task": self.name}
        self.run_id = self._client.record_run_start(
            end_user_id=self.end_user_id,
            session_id=self.session_id,
            feature_tag=self.feature_tag,
            deployment_version=self.deployment_version,
            metadata=merged_metadata,
            intent=self.intent,
        )
        self._ctx = self._client.context(
            run_id=self.run_id,
            end_user_id=self.end_user_id,
            session_id=self.session_id,
            feature_tag=self.feature_tag,
            deployment_version=self.deployment_version,
        )
        self._ctx.__enter__()
        return self

    def __exit__(self, exc_type: Any, exc: BaseException | None, tb: Any) -> None:
        if self._ctx is not None:
            self._ctx.__exit__(exc_type, exc, tb)
        if self._outcome_recorded or self.run_id is None:
            return
        if exc is not None:
            self.outcome(
                "failed",
                success=False,
                final_status="failed",
                labels={"error_type": exc.__class__.__name__, "task": self.name},
            )
            return
        self.outcome(
            "completed",
            success=True,
            labels={"task": self.name},
        )

    def span(
        self,
        name: str,
        *,
        span_type: str = "agent",
        status: str = "succeeded",
        cost_usd: Decimal | float | str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> str:
        return self._client.record_span(
            run_id=self._require_run_id(),
            name=name,
            span_type=span_type,
            status=status,
            cost_usd=cost_usd,
            metadata=metadata,
        )

    def tool_call(
        self,
        tool_name: str,
        *,
        span_id: str | None = None,
        tool_type: str = "read",
        risk_score: int | None = None,
        duration_ms: int | None = None,
        status: str = "success",
    ) -> None:
        self._client.record_tool_call(
            run_id=self._require_run_id(),
            span_id=span_id,
            tool_name=tool_name,
            tool_type=tool_type,
            risk_score=risk_score,
            duration_ms=duration_ms,
            status=status,
        )

    def model_call(
        self,
        *,
        provider: str,
        model: str,
        span_id: str | None = None,
        input_tokens: int | None = None,
        output_tokens: int | None = None,
        cached_input_tokens: int | None = None,
        latency_ms: int | None = None,
        cost_usd: Decimal | float | str | None = None,
        status: str = "success",
        error_type: str | None = None,
    ) -> None:
        self._client.record_model_call(
            run_id=self._require_run_id(),
            provider=provider,
            model=model,
            span_id=span_id,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cached_input_tokens=cached_input_tokens,
            latency_ms=latency_ms,
            cost_usd=cost_usd,
            status=status,
            error_type=error_type,
        )
        if cost_usd is not None:
            self._total_cost_usd += Decimal(str(cost_usd))
        self._total_input_tokens += int(input_tokens or 0)
        self._total_output_tokens += int(output_tokens or 0)

    def outcome(
        self,
        outcome_type: str,
        *,
        success: bool = True,
        labels: dict[str, Any] | None = None,
        final_status: str = "succeeded",
        total_cost_usd: Decimal | float | str | None = None,
        total_input_tokens: int | None = None,
        total_output_tokens: int | None = None,
        quality_score: float | None = None,
        verification_status: str | None = None,
    ) -> None:
        merged_labels = dict(labels or {})
        if quality_score is not None:
            merged_labels["quality_score"] = quality_score
        if verification_status is not None:
            merged_labels["verification_status"] = verification_status
        self._client.record_outcome(
            run_id=self._require_run_id(),
            outcome_type=outcome_type,
            success=success,
            labels=merged_labels or None,
            final_status=final_status,
            total_cost_usd=total_cost_usd
            if total_cost_usd is not None
            else str(self._total_cost_usd),
            total_input_tokens=total_input_tokens
            if total_input_tokens is not None
            else self._total_input_tokens,
            total_output_tokens=total_output_tokens
            if total_output_tokens is not None
            else self._total_output_tokens,
        )
        self._outcome_recorded = True

    def check_budget(self, **kwargs: Any) -> dict[str, Any]:
        return self._client.check_budget(**kwargs)

    def check_policy(self, **kwargs: Any) -> dict[str, Any]:
        return self._client.check_policy(**kwargs)

    def _require_run_id(self) -> str:
        if self.run_id is None:
            raise RuntimeError("RunLedgerTask has not been entered yet")
        return self.run_id
